### ISSUE TITLE
ValueError: Length mismatch: Expected axis has 0 elements, new values have 10 elements

### ISSUE THREAD

**Issue by nrminor** [CFIA-NCFAD/wgscovplot#49]
Hello,

Our team has been working on running wgscovplot on results from a recent viralrecon run. However, it keeps failing with what I believe is a pandas empty-dataframe error, potentially at the point where it tries to summarize coverage.

I've pasted the final portion of the traceback below (sorry for the mangled rich-formatting!), but the whole output from the command is attached.
```python
│ /Users/redacted/opt/miniconda3/envs/wgscovplot/lib/python3.9/site-packages/pandas/core/internals/base │
│ .py:98 in _validate_set_axis                                                                     │
│                                                                                                  │
│    95 │   │   │   pass                                                                           │
│    96 │   │                                                                                      │
│    97 │   │   elif new_len != old_len:                                                           │
│ ❱  98 │   │   │   raise ValueError(                                                              │
│    99 │   │   │   │   f"Length mismatch: Expected axis has {old_len} elements, new "             │
│   100 │   │   │   │   f"values have {new_len} elements"                                          │
│   101 │   │   │   )                                                                              │
│                                                                                                  │
│ ╭───────────────────────────────────── locals ─────────────────────────────────────╮             │
│ │       axis = 0                                                                   │             │
│ │ new_labels = Index(['Sample', '# 0 Coverage Positions', '0 Coverage Regions',    │             │
│ │              │      'Low Coverage Threshold', '# Low Coverage Positions (< 2X)', │             │
│ │              │      'Low Coverage Regions (< 2X)', '% Genome Coverage >= 2X',    │             │
│ │              │      'Mean Coverage Depth (X)', 'Median Coverage Depth (X)',      │             │
│ │              │      'Ref Sequence Length (bp)'],                                 │             │
│ │              │     dtype='object')                                               │             │
│ │    new_len = 10                                                                  │             │
│ │    old_len = 0                                                                   │             │
│ │       self = BlockManager                                                        │             │
│ │              Items: RangeIndex(start=0, stop=0, step=1)                          │             │
│ │              Axis 1: RangeIndex(start=0, stop=0, step=1)                         │             │
│ ╰──────────────────────────────────────────────────────────────────────────────────╯             │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ValueError: Length mismatch: Expected axis has 0 elements, new values have 10 elements
```

Would you have any advice on how to fix this? Could it be that it's looking for a file that is now placed in a different location after a couple years of viralrecon updates?

Thanks in advance,
--Nick

[output.txt](https://github.com/nhhaidee/wgscovplot/files/14223371/output.txt)

**Comment by nhhaidee** [comment]
Can you check if mosdepth files are in the resutl folder? The tool will find these files as below

`
    '**/mosdepth/**/*.genome.per-base.bed.gz',
    '**/mosdepth/**/*.per-base.bed.gz',
`

**Comment by nrminor** [comment]
We do have a mosdepth folder in the results, but there aren't any files matching those globs. Would it be helpful if I shared a list of the files that are there?

**Comment by nhhaidee** [comment]
Currently, the tool will locate mosdepth file from folder with pattern that I mentioned. Can you share directory path to mosdepth file? With viralrecon we can specify where mosdepth files can be saved in result folder!

**Comment by nhhaidee** [comment]
What is your mosdepth file patterns? We can improve the tool in the future!

**Comment by nhhaidee** [comment]
Oh! It seems viralrecon does not save mosdepth file `*.per-base.bed.gz`. I have not run viralrecon for a while, but when I ran it I can specify to save `*.per-base.bed.gz` file with below format.

@peterk87 Do you have comment about this? Is there any updates from viralrecon

MN908947.3	0	34	0
MN908947.3	34	75	387
MN908947.3	75	94	341
MN908947.3	94	110	1001
MN908947.3	110	183	660
MN908947.3	183	191	743
MN908947.3	191	220	87

**Comment by peterk87** [comment]
I've usually had to run viralrecon with a custom conf so that the Mosdepth output is published (see [custom conf here](https://gist.github.com/peterk87/994cff3f0364aa6051feacf76dbf2568); hopefully still works with the latest versions of viralrecon). By default, only the summary txt is output (see [here](https://github.com/nf-core/viralrecon/blob/3731dd3a32a67a2648ea22c2bd980c224abdaee2/conf/modules_illumina.config#L338))

The next version of wgscovplot will try to look for Mosdepth per-base.bed.gz files, and if not found, it will look for BAM files and try to compute the depths from those using pysam. This will be similar to what another one of our tools is does (see [xlavir parse_bam_depths](https://github.com/CFIA-NCFAD/xlavir/blob/2dce2923450f57982191ebef016c4cb7bdb5e4ea/xlavir/tools/mosdepth.py#L116)).

**Comment by nrminor** [comment]
Well there we go! Thanks for sharing that config @peterk87. We're already using custom configs to tweak some other settings, so we'll just tack on what you had in yours.

**Comment by peterk87** [comment]
Sounds good! Let us know if you have any other issues!

**Comment by nrminor** [comment]
Will do! Thanks so much for both of your prompt assistance here.

**Comment by peterk87** [comment]
Parsing of BAM files as a fallback was implemented in #50 and release [1.0.0](https://github.com/CFIA-NCFAD/wgscovplot/releases/tag/1.0.0). Unfortunately, it's a slow especially compared to simply reading Mosdepth `per-base.bed.gz` files, which Mosdepth can generate much more quickly than Python + pysam can. 

I'm closing this issue as complete. But if it's still an issue, please let us know and we'll reopen and try to address it once and for all!

**Comment by nrminor** [comment]
All good! Thanks again Peter!
